gossipd/queries.c

```diff
@@ -485,89 +528,108 @@ static bool queue_channel_ranges(struct peer *peer,
+			continue;
 
 		get_checksum_and_timestamp(rstate, chan, 0,
 					   &ts.timestamp_node_id_1,
 					   &cs.checksum_node_id_1);
 		get_checksum_and_timestamp(rstate, chan, 1,
 					   &ts.timestamp_node_id_2,
 					   &cs.checksum_node_id_2);
-
-		if (csums)
-			tal_arr_expand(&csums, cs);
-		if (tstamps)
-			encoding_add_timestamps(&tstamps->encoded_timestamps,
-						&ts);
+		if (query_option_flags & QUERY_ADD_TIMESTAMPS)
+			tal_arr_expand(tstamps, ts);
+		if (query_option_flags & QUERY_ADD_CHECKSUMS)
+			tal_arr_expand(csums, cs);
 	}
 
-	extension_bytes = 0;
-
-	/* If either of these can't fit in max_encoded_bytes by itself,
-	 * it's over. */
-	if (csums) {
-		extension_bytes += tlv_len(csums);
-	}
-
-	if (tstamps) {
-		if (!encoding_end_external_type(&tstamps->encoded_timestamps,
-						&tstamps->encoding_type,
-						max_encoded_bytes))
-			goto wont_fit;
-		/* 1 byte for encoding_type, too */
-		extension_bytes += 1 + tlv_len(tstamps->encoded_timestamps);
-	}
-
-	/* If we can encode that, fine: send it */
-	if (extension_bytes <= max_encoded_bytes
-	    && encoding_end_prepend_type(&encoded_scids,
-					 max_encoded_bytes - extension_bytes)) {
-		reply_channel_range(peer, first_blocknum,
-				    number_of_blocks + tail_blocks,
-				    encoded_scids,
-				    tstamps, csums);
-		return true;
-	}
+	return scids;
+}
 
-wont_fit:
-	/* It wouldn't all fit: divide in half */
-	/* We assume we can always send one block! */
-	if (number_of_blocks <= 1) {
-		/* We always assume we can send 1 blocks worth */
-		status_broken("Could not fit scids for single block %u",
-			      first_blocknum);
-		return false;
-	}
-	status_debug("queue_channel_ranges full: splitting %u+%u and %u+%u(+%u)",
-		     first_blocknum,
-		     number_of_blocks / 2,
-		     first_blocknum + number_of_blocks / 2,
-		     number_of_blocks - number_of_blocks / 2,
-		     tail_blocks);
-	return queue_channel_ranges(peer, first_blocknum, number_of_blocks / 2,
-				    0, query_option_flags)
-		&& queue_channel_ranges(peer, first_blocknum + number_of_blocks / 2,
-					number_of_blocks - number_of_blocks / 2,
-					tail_blocks, query_option_flags);
+/*~ When we need to send an array of channels, it might go over our 64k packet
+ * size.  But because we use compression, we can't actually tell how much
+ * we'll use.  We pack them into the maximum amount for uncompressed, then
+ * compress afterwards.
+ */
+static void queue_channel_ranges(struct peer *peer,
+				 u32 first_blocknum, u32 number_of_blocks,
+				 enum query_option_flags query_option_flags)
+{
+	struct routing_state *rstate = peer->daemon->rstate;
+	struct channel_update_timestamps *tstamps;
+	struct channel_update_checksums *csums;
+	struct short_channel_id *scids;
+	size_t off, limit;
+
+	scids = gather_range(tmpctx, rstate, first_blocknum, number_of_blocks,
+			     query_option_flags, &tstamps, &csums);
+
+	limit = max_entries(query_option_flags);
+	off = 0;
```

Comment:
Throughout the rest of this function the offset variable `size_t off` is untouched and will always be constant `0`. It will only used for adding or subtracting on some integer or indexes. I think this is a leftover of some prior code state of yours ;)

Reply:
I assume it's for the splitter logic where a single block can't fit the reply, and therefore the later replies will start from `off`. But as @m-schmoock points out it isn't set anywhere, so presumably it should be set to `n` on line 605, before decrementing `n--` to make it fit :-)

Reply:
Good catch, fixed.